src/os/bluestore/BlueFSCache.cc

```diff
@@ -0,0 +1,203 @@
+#include "BlueFSCache.h"
+#include "bluestore_common.h"
+#include <cerrno>
+#include <cstddef>
+#include <cstdint>
+#include <functional>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <shared_mutex>
+#include <unordered_map>
+#include "common/errno.h"
+#include "common/dout.h"
+#include "common/debug.h"
+#include "include/buffer_fwd.h"
+#include "common/ceph_context.h"
+#include "include/denc.h"
+#include "tools/cephfs/EventOutput.h"
+
+
+#define dout_context cct
+#define dout_subsys ceph_subsys_bluefs
+#undef dout_prefix
+#define dout_prefix *_dout << "bluefs "
+
+LRUCache::LRUCache(size_t cache_capacity, size_t evict_bytes, CephContext* cct)
+        : cache_capacity(cache_capacity), evict_bytes(evict_bytes), current_bytes(0), cct(cct) {}
+
+cache_bufferlist LRUCache::get_nearest_offset(uint64_t inode, uint64_t offset,
+                                              uint64_t *buffer_offset) {
+
+  dout(20) << __func__ << " inode " << inode << std::hex << " offset 0x" << offset << std::dec << dendl;
+  std::shared_lock lock(mutex);
+  inode_offset_pair key = {inode, offset};
+  auto it = kv.find(key);
+  if (it != kv.end()) {
+    dout(20) << __func__ << " inode " << inode << std::hex << " offset 0x" << offset << " exists in cache" << std::dec << dendl;
+    lock.unlock();
+    *buffer_offset = 0;
+    update_recency(key);
+    return it->second;
+  }
+  it = kv.upper_bound({inode, offset});
```

Comment:
How do you know that ordering of std::pair is that
for x0 < x1 => (x0, ?) < (x1, ?)
and
for y0 < y1 => (x, y0) < (x, y1)
?
Where can I read about it?

Reply:
@aclamk, is this what you're after https://en.cppreference.com/w/cpp/utility/pair/operator_cmp.html ?